File: spring-reactive-02/learn-java/src/main/java/boki/learnjava/v7_split/c_groupby/GroupByEx1.java
package boki.learnjava.v7_split.c_groupby;

import boki.learnjava.common.SampleData;
import boki.learnjava.utils.Logger;
import reactor.core.publisher.Flux;

/**
 * groupBy(keyMapper) 기본 개념 예제
 *  - emit되는 데이터를 keyMapper로 생성한 key를 기준으로 그룹화 한 GroupedFlux를 리턴한다.
 */
public class GroupByEx1 {
    public static void main(String[] args) {
        Flux
            .fromIterable(SampleData.books)
            .groupBy(book -> book.getAuthorName())
            .flatMap(groupedFlux ->
                groupedFlux
                    .map(book -> book.getBookName() +
                        "(" + book.getAuthorName() + ")")
                    .collectList()
            )
            .subscribe(bookByAuthor ->
                Logger.onNext(bookByAuthor));
    }
}

File: spring-reactive-02/learn-kt/src/main/kotlin/boki/learnkt/v7_split/c_groupby/GroupByEx1.kt
package boki.learnkt.v7_split.c_groupby

import boki.learnjava.common.SampleData
import boki.learnkt.util.Logger
import reactor.core.publisher.Flux

/**
 * groupBy(keyMapper) 기본 개념 예제
 *  - emit되는 데이터를 keyMapper로 생성한 key를 기준으로 그룹화 한 GroupedFlux를 리턴한다.
 */
fun main() {
    Flux
        .fromIterable(SampleData.books)
        .groupBy { it.authorName }
        .flatMap { book -> book.map { "${it.bookName}(${it.authorName})" }.collectList() }
        .subscribe { Logger.onNext(it) }
}